File: backend/app/schemas/knowledge.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel


class KnowledgeItemRead(BaseModel):
    id: UUID
    organization_id: UUID
    project_id: UUID
    source: str
    source_entity_id: str
    entity_type: str
    title: str
    description: str | None = None
    url: str | None = None
    author: str | None = None
    created_time: datetime
    updated_time: datetime
    tags: list[str] | None = None
    status: str

    model_config = {"from_attributes": True}


class KnowledgeItemDetail(KnowledgeItemRead):
    content: str | None = None
    metadata_json: dict | None = None

    model_config = {"from_attributes": True}


class KnowledgeRelationshipRead(BaseModel):
    id: UUID
    source_item_id: UUID
    target_item_id: UUID
    relationship_type: str
    confidence: float
    created_at: datetime

    # We will include basic properties of source/target items for detail rendering
    source_item: KnowledgeItemRead | None = None
    target_item: KnowledgeItemRead | None = None

    model_config = {"from_attributes": True}


class KnowledgeItemListResponse(BaseModel):
    items: list[KnowledgeItemRead]
    total: int


class KnowledgeSyncLogRead(BaseModel):
    id: UUID
    project_id: UUID
    started_at: datetime
    completed_at: datetime | None = None
    status: str
    items_synced: int
    relationships_created: int
    error_message: str | None = None

    model_config = {"from_attributes": True}

File: frontend/src/types/knowledge.ts
export interface KnowledgeItem {
  id: string;
  organization_id: string;
  project_id: string;
  source: "github" | "notion" | "google_drive" | string;
  source_entity_id: string;
  entity_type: "commit" | "pull_request" | "issue" | "notion_page" | "document" | "folder" | string;
  title: string;
  description: string | null;
  content?: string | null;
  url: string | null;
  author: string | null;
  created_time: string;
  updated_time: string;
  tags: string[] | null;
  status: string;
  metadata_json?: Record<string, any> | null;
}

export interface KnowledgeRelationship {
  id: string;
  source_item_id: string;
  target_item_id: string;
  relationship_type: string;
  confidence: number;
  created_at: string;
  source_item?: KnowledgeItem;
  target_item?: KnowledgeItem;
}

export interface KnowledgeSyncLog {
  id: string;
  project_id: string;
  started_at: string;
  completed_at: string | null;
  status: "running" | "completed" | "failed";
  items_synced: number;
  relationships_created: number;
  error_message: string | null;
}

export interface KnowledgeItemListResponse {
  items: KnowledgeItem[];
  total: number;
}
